File: Embarcadero/Rio/CPP/CustomListBox/customlistfrm.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "customlistfrm.h"
#include "DelphiIntf.hpp"
// ---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"

long rndm(long max) {
#if  defined __APPLE__ || defined __ANDROID__
	return (random() % ((max + 1) + max));
#else
	return random(max);
#endif
}

TfrmCustomList *frmCustomList;

// ---------------------------------------------------------------------------
TFmxObject *FindItemParent(TFmxObject* Obj, String ParentClass) {
	TFmxObject* result;
	if (Obj->Parent != NULL) {
		if (SameText(Obj->Parent->ClassName(), ParentClass)) {
			return Obj->Parent;
		} else {
			return FindItemParent(Obj->Parent, ParentClass);
		}
	}
	return NULL;
}
// ---------------------------------------------------------------------------
__fastcall TfrmCustomList::TfrmCustomList(TComponent* Owner) : TForm(Owner) {
	OpenDialog1->Filter = TBitmapCodecManager::GetFilterString();
}

// ---------------------------------------------------------------------------
void __fastcall TfrmCustomList::DoInfoClick(TObject *Sender) {
	TListBoxItem * Item = (TListBoxItem *) FindItemParent((TFmxObject *)Sender, "TListBoxItem");
	InfoLabel->Text = "Info Button click on " + IntToStr(Item->Index) +
		" listbox item";
}
// ---------------------------------------------------------------------------
void __fastcall TfrmCustomList::DoVisibleChange(TObject *Sender) {
	TListBoxItem * Item = (TListBoxItem *) FindItemParent((TFmxObject *)Sender, "TListBoxItem");
	InfoLabel->Text = "Checkbox changed " + IntToStr(ListBox1->ItemIndex) +
		" listbox item to " + BoolToStr
		(Item->StylesData["visible"].AsBoolean(), true);
}

// ---------------------------------------------------------------------------
void __fastcall TfrmCustomList::Button1Click(TObject *Sender) {
	TListBoxItem *Item;
	int i;

	if (OpenDialog1->Execute()) {
		for (i = 0; i < OpenDialog1->Files->Count; i++) {
			String currentFile = OpenDialog1->Files->operator [](i);
			Item = new TListBoxItem(NULL);
			Item->Parent = ListBox1;
			Item->StyleLookup = "CustomItem";
			Item->TagString = currentFile;
			Item->Text = currentFile;
			Item->StylesData["icon"] = TValue::From<String>(currentFile);
			Item->StylesData["resolution"] = TValue::From<String>("1024x768 px"); // set size
			Item->StylesData["depth"] = TValue::From<String>("32 bit");
			Item->StylesData["visible"] = TValue::From<bool>(true); // set Checkbox value
			Item->StylesData["visible.OnChange"] = NotifyEventAsTValue(DoVisibleChange); // set OnChange value
			Item->StylesData["info.OnClick"] = NotifyEventAsTValue(DoInfoClick); // set OnClick value
		}
		Caption = IntToStr(ListBox1->Count) + " items";
	};
}

// ---------------------------------------------------------------------------
void __fastcall TfrmCustomList::Button2Click(TObject *Sender) {
	TListBoxItem *Item;

	Item = new TListBoxItem(NULL);
	Item->Parent = ListBox1;
	Item->StyleLookup = "CustomItem";
	Item->Text = "item " + IntToStr(Item->Index);
	if (Item->Index % 2) {
		Item->ItemData->Bitmap = Image1->Bitmap;
	} else {
		Item->ItemData->Bitmap = Image2->Bitmap;
	}
	Item->StylesData["resolution"] = TValue::From<String>("1024x768 px"); // set size
	Item->StylesData["depth"] = TValue::From<String>("32 bit");
	Item->StylesData["visible"] = TValue::From<bool>(true); // set Checkbox value
	Item->StylesData["visible.OnChange"] = NotifyEventAsTValue(DoVisibleChange); // set OnChange value
	Item->StylesData["info.OnClick"] = NotifyEventAsTValue(DoInfoClick); // set OnClick value
   	Caption = IntToStr(ListBox1->Count) + " items";
}
// ---------------------------------------------------------------------------

void __fastcall TfrmCustomList::CheckBox1Change(TObject *Sender) {
	ListBox1->AllowDrag = CheckBox1->IsChecked;
}
// ---------------------------------------------------------------------------

void __fastcall TfrmCustomList::Button3Click(TObject *Sender) {
	int i;
	ListBox1->BeginUpdate();
	for (i = 1; i < 1000; i++)
		Button2Click(Sender);
	ListBox1->EndUpdate();
}
// ---------------------------------------------------------------------------

File: Embarcadero/Rio/Object Pascal/CustomListBox/customlistfrm.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit customlistfrm;

interface

uses
  System.SysUtils, System.Variants, System.Classes, System.Types, System.UITypes,
  System.Rtti, FMX.Forms, FMX.Dialogs, FMX.Types, FMX.Layouts, FMX.Styles, FMX.StdCtrls,
  FMX.ListBox, FMX.Objects, FMX.Controls, FMX.Edit, FMX.Effects, FMX.Graphics;

type
  TfrmCustomList = class(TForm)
    ListBox1: TListBox;
    Resources1: TStyleBook;
    OpenDialog1: TOpenDialog;
    InfoLabel: TLabel;
    Label1: TLabel;
    Button2: TButton;
    Button3: TButton;
    CheckBox1: TCheckBox;
    Image1: TImage;
    Image2: TImage;
    Button1: TButton;
    procedure Button1Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure Button3Click(Sender: TObject);
    procedure CheckBox1Change(Sender: TObject);
  private
    { Private declarations }
    procedure DoInfoClick(Sender: TObject);
    procedure DoVisibleChange(Sender: TObject);
  public
    { Public declarations }
  end;

var
  frmCustomList: TfrmCustomList;

implementation

{$R *.fmx}
{$R *.Windows.fmx MSWINDOWS}

function FindItemParent(Obj: TFmxObject; ParentClass: TClass): TFmxObject;
begin
  Result := nil;
  if Assigned(Obj.Parent) then
    if Obj.Parent.ClassType = ParentClass then
      Result := Obj.Parent
    else
      Result := FindItemParent(Obj.Parent, ParentClass);
end;


procedure TfrmCustomList.Button1Click(Sender: TObject);
var
  Item: TListBoxItem;
  I: Integer;
begin
  OpenDialog1.Filter := TBitmapCodecManager.GetFilterString;
  if OpenDialog1.Execute then
  begin
    // create item and save file name in the tag
    for I := 0 to OpenDialog1.Files.Count - 1 do
    begin
      Item := TListBoxItem.Create(nil);
      Item.Parent := ListBox1;
      Item.TagString := OpenDialog1.Files[I];
      Item.StyleLookup := 'CustomItem';
      Item.Text := OpenDialog1.Files[i]; // set filename
      Item.StylesData['icon'] := OpenDialog1.Files[i];
      Item.StylesData['resolution'] := '1024x768 px'; // set size
      Item.StylesData['depth'] := '32 bit';
      Item.StylesData['visible'] := true; // set Checkbox value
      Item.StylesData['visible.OnChange'] := TValue.From<TNotifyEvent>(DoVisibleChange); // set OnChange value
      Item.StylesData['info.OnClick'] := TValue.From<TNotifyEvent>(DoInfoClick); // set OnClick value
    end;
    Caption := IntToStr(ListBox1.Count) + ' items';
  end;
end;

procedure TfrmCustomList.Button2Click(Sender: TObject);
var
  Item: TListBoxItem;
begin
  // create custom item
  Item := TListBoxItem.Create(nil);
  Item.Parent := ListBox1;
  Item.StyleLookup := 'CustomItem';
  Item.Text := 'item ' + IntToStr(Item.Index); // set filename
  if Odd(Item.Index) then
    Item.ItemData.Bitmap := Image1.Bitmap // set thumbnail
  else
    Item.ItemData.Bitmap := Image2.Bitmap; // set thumbnail
  Item.StylesData['resolution'] := '1024x768 px'; // set size
  Item.StylesData['depth'] := '32 bit';
  Item.StylesData['visible'] := true; // set Checkbox value
  Item.StylesData['visible.OnChange'] := TValue.From<TNotifyEvent>(DoVisibleChange); // set OnChange value
  Item.StylesData['info.OnClick'] := TValue.From<TNotifyEvent>(DoInfoClick); // set OnClick value
end;

procedure TfrmCustomList.DoInfoClick(Sender: TObject);
var
  Item : TListBoxItem;
begin
  Item := TListBoxItem(FindItemParent(Sender as TFmxObject,TListBoxItem));
  if Assigned(Item) then
    InfoLabel.Text := 'Info Button click on ' + IntToStr(Item.Index) + ' listbox item';
end;

procedure TfrmCustomList.DoVisibleChange(Sender: TObject);
var
  Item : TListBoxItem;
begin
  Item := TListBoxItem(FindItemParent(Sender as TFmxObject,TListBoxItem));
  if Assigned(Item) then
    InfoLabel.Text := 'Checkbox changed ' + IntToStr(Item.Index) + ' listbox item to ' + BoolToStr(Item.StylesData['visible'].AsBoolean, true);
end;

procedure TfrmCustomList.Button3Click(Sender: TObject);
var
  i: integer;
begin
  ListBox1.BeginUpdate;
  for i := 1 to 1000 do
    Button2Click(Sender);
  ListBox1.EndUpdate;

  Caption := IntToStr(ListBox1.Count) + ' items';
end;

procedure TfrmCustomList.CheckBox1Change(Sender: TObject);
begin
  ListBox1.AllowDrag := CheckBox1.IsChecked;
end;

end.
